src/main/java/net/ripe/rpki/rsyncit/rrdp/State.java

```diff
@@ -25,14 +29,15 @@ public State() {
     }
 
     // Remove entries that were not mentioned in RRDP repository for a while
-    public synchronized void removeOldObject(Instant cutOffTime) {
-        var hashesToDelete = new ArrayList<>();
-        times.forEach((hash, t) -> {
-            if (t.getLastMentioned().isBefore(cutOffTime)) {
-                hashesToDelete.add(hash);
-            }
-        });
-        hashesToDelete.forEach(times::remove);
+    public void removeOldObject(Instant cutOffTime) {
+        var expired = times.entrySet().stream().filter(entry ->
+            entry.getValue().getLastMentioned().isBefore(cutOffTime)
+        );
+
+        var removed = expired.mapToInt(entry -> Boolean.compare(times.remove(entry.getKey(), entry.getValue()), false));
```

Comment:
`var removedCount = expired.filter(entry -> times.remove(entry.getKey(), entry.getValue())).count();`?

Reply:
I find this change much harder to understand, to be honest.
The simplest change would be to log `hashesToDelete.size()`

Reply:
My assumption is that the `remove` can fail sometimes because of concurrent modification.

Reply:
Yeap, that's why it was synchronized

Reply:
`synchronised` here would just protect against concurrent remove calls. Not against cache modification while running.

@erikrozendaal nice usage of filter 👍

Reply:
Cache modification while running are not a problem here, it doesn't matter much.
But, as I said, I don't like this change since it
- needs more time to figure out what's going on
- requires the knowledge that iteration over `expired` is not influences by `times.remove`

Reply:
> needs more time to figure out what's going on

I do not think a performance change by using the "safe" version of remove (that checks that the value has not been mutated) will be significant.

> requires the knowledge that iteration over expired is not influenced/broken by times.remove

Fair point. I thought I used a list as an intermediate result but did not. Added.

Reply:
> I do not think a performance change by using the "safe" version of remove (that checks that the value has not been mutated) will be significant.

I'm talking about figuring out what the code does by looking at it, not performance.